### PR TITLE
base class to check for discouraged clients

### PR DESCRIPTION
**Pull request by Njeriiii**
What does this implement/fix? Explain your changes.
---------------------------------------------------
This class extends the LocalInspectionTool to check for the use of discouraged clients in the code and suggests using other clients instead.
 The client data is loaded from the configuration file and the client name is checked against the
 discouraged client name. If the client name matches, a problem is registered with the suggestion message.

#### b. Use **`EventProcessorClient`** instead of **`EventHubConsumerAsyncClient`**

PR https://github.com/Njeriiii/azure-tools-for-java/pull/5 Is refactored to make a base class that loads clients to check and makes suggestions

Has this been tested?
---------------------------
- [x] Tested

Any relevant logs, screenshots, error output, etc.?
-------------------------------------
Screenshot 2024-07-02 182801

Screenshot 2024-07-02 182730

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/.idea/misc.xml
Comment (Njeriiii) [inline PluginsAndFeatures/azure-toolkit-for-intellij/.idea/misc.xml:1]: this .idea file keeps getting committed(in commits in other PRs) -- I just added it to gitignore above not sure why its still getting committed. (there's 2 .idea directories and the outer-level one is the one whose misc.xml file has been tracked

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/README.md [hunk 1]
```diff
@@ -16,16 +16,32 @@ The Java Code Quality Analyzer, is a plugin designed to improve the quality of J
...
 - **Issue**: This can lead to OutOfMemoryErrors, especially with large files or high-volume uploads.
 - **Severity: INFO**
 - **Recommendation**: Use APIs that take a length parameter. Please refer to the [Azure SDK for Java documentation](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-blob-upload-java) for additional information.
 
-2. #### Use ServiceBusProcessorClient instead of ServiceBusReceiverAsyncClient.
-- **Anti-pattern**: The use of the Reactor receiver, specifically the `ServiceBusReceiverAsyncClient`, is an anti-pattern. This is because it's a low-level API that provides fine-grained control over message handling. While this might seem beneficial, it requires a high level of proficiency in Reactive programming and is mainly useful when building a Reactive library or an end-to-end Reactive application.
 
-- **Issue**: The main issue with using `ServiceBusReceiverAsyncClient` is its complexity and the requirement for a deep understanding of Reactive programming. This can make it difficult to use correctly and efficiently, especially for developers who are not familiar with Reactive programming paradigms.
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/README.md]: Use sentence casing.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/azure-intellij-plugin-azure-sdk.xml [hunk 2]
```diff
@@ -40,15 +40,15 @@
 
         <localInspection
                 language="JAVA"
-                shortName="ServiceBusReceiverAsyncClientCheck"
-                displayName="Low level ServiceBusReceiverAsyncClient check"
+                shortName="DetectDiscouragedClientCheck"
+                displayName="Detect Discouraged Client Check check"
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/azure-intellij-plugin-azure-sdk.xml]: casing update

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json [hunk 3]
```diff
@@ -6,9 +6,14 @@
...
-    "client_name": "ServiceBusReceiverAsyncClient",
-    "antipattern_message": "Use of ServiceBusReceiverAsyncClient detected. Use ServiceBusProcessorClient instead."
+  "DetectDiscouragedClientCheck": {
+    "ServiceBusReceiverAsyncClientCheck": {
+      "client_name": "ServiceBusReceiverAsyncClient",
+      "antipattern_message": "Use of ServiceBusReceiverAsyncClient detected. Use ServiceBusProcessorClient instead."
+    },
+    "EventHubConsumerAsyncClientCheck": {
+      "client_name": "EventHubConsumerAsyncClient",
+      "antipattern_message": "Use of EventHubConsumerAsyncClient detected. Use EventProcessorClient instead."
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json]: We should make the message more informative, to include a few things on the lines
>provides a higher-level abstraction that simplifies event processing, making it the preferred choice for most developers.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DetectDiscouragedClientCheck.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+public class DetectDiscouragedClientCheck extends LocalInspectionTool {
+
+    @NotNull
+    @Override
+    public JavaElementVisitor buildVisitor(@NotNull ProblemsHolder holder,boolean isOnTheFly){
+        return new DetectDiscouragedClientVisitor(holder,isOnTheFly);
+    }
+
+    public static class DetectDiscouragedClientVisitor extends JavaElementVisitor {
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DetectDiscouragedClientCheck.java]: do we need this to be public?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json [hunk 5]
```diff
@@ -51,10 +51,13 @@
...
     ],
-    "methods_to_check": ["receiveMode","prefetchCount"],
+    "methods_to_check": [
+      "receiveMode",
+      "prefetchCount"
+    ],
     "anti_pattern_message": "A high prefetch value in PEEK_LOCK detected. We recommend a prefetch value of 0 or 1 for efficient message retrieval."
   },
-  "DetectDiscouragedAPIUsageCheck" : {
+  "DetectDiscouragedAPIUsageCheck": {
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json:60]: Isn't the actual rule definition on L71?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/azure-intellij-plugin-azure-sdk.xml [hunk 6]
```diff
@@ -81,16 +67,16 @@
...
+                shortName="HardcodedAPIKeysAndTokensCheck"
+                displayName="Hardcoded API Keys And Tokens Check"
+                groupName="Azure SDK Rules"
+                enabledByDefault="true"
+                level="WARNING"
+                implementationClass="com.microsoft.azure.toolkit.intellij.azure.sdk.buildtool.HardcodedAPIKeysAndTokensCheck">
 
             <class>
-            com.microsoft.azure.toolkit.intellij.azure.sdk.buildtool.HardcodedAPIKeysAndTokensCheck
+                com.microsoft.azure.toolkit.intellij.azure.sdk.buildtool.HardcodedAPIKeysAndTokensCheck
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/azure-intellij-plugin-azure-sdk.xml]: revert?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/RuleConfigLoader.java [hunk 7]
```diff
@@ -209,6 +217,7 @@ private Map<String, String> getMapOfDiscouragedIdentifiers(JsonReader reader, Ma
 
             switch (fieldName) {
                 case "methods_to_check":
+                case "clients_to_check":
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/RuleConfigLoader.java:220]: nit: Do we do the same handling for methods_to_check vs client to check?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DetectDiscouragedClientCheck.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        public DetectDiscouragedClientVisitor(ProblemsHolder holder, boolean isOnTheFly) {
+            this.holder = holder;
+            this.isOnTheFly = isOnTheFly;
+        }
+
+        private static Map<String, String> CLIENT_DATA;
+
+        // Define constants for string literals
+        private static final RuleConfig RULE_CONFIG;
+        private static final Map<String, String> CLIENTS_TO_CHECK;
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DetectDiscouragedClientCheck.java:51]: needs to be updated with main json loader changes

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/README.md [hunk 9]
```diff
@@ -159,3 +159,43 @@ integration, telemetry connectivity, and Azure Toolkit integration.
...
+    ##### a. **ServiceBusReceiverAsyncClient**
+
+   - **Anti-pattern**: The `ServiceBusReceiverAsyncClient` is considered an anti-pattern because it demands detailed
+   handling of messages, which can be overly complex and unnecessary for most common use cases.
+   - **Severity: WARNING**
+   - **Recommendation**: Instead of using `ServiceBusReceiverAsyncClient`, it is recommended to
+     use `ServiceBusProcessorClient`. The `ServiceBusProcessorClient` is a higher-level abstraction that simplifies
+     message consumption, making it a more suitable option for most developers and scenarios.
+   - Please refer to
+     the [Azure Service Bus client library for Java](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/azure-messaging-servicebus/README.md#when-to-use-servicebusprocessorclient)
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/README.md]: nit: refer to Service Bus client ?

**Review by samvaity — APPROVED** [review]

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json (hunk 5 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json:60]: there's discouraged APIs and discouraged clients

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/RuleConfigLoader.java (hunk 7 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/RuleConfigLoader.java:220]: yes they'll both be keys in the map

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DetectDiscouragedClientCheck.java (hunk 8 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DetectDiscouragedClientCheck.java:51]: this is up to date -- unless there's something I'm missing?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/azure-intellij-plugin-azure-sdk.xml (hunk 6 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/azure-intellij-plugin-azure-sdk.xml]: done:)

**Pull request merged**